### PR TITLE
Deprecate `LightningModule.on_pretrain_routine_{start/end}`

### LINKED ISSUE #10984: [RFC] Deprecate `on_pretrain_routine_start` and `on_pretrain_routine_end` LM/callback hooks
## Proposed refactor

Deprecate these hooks:
https://github.com/PyTorchLightning/pytorch-lightning/blob/6369e3b77fa3f38613b661517f6361f842f611c9/pytorch_lightning/trainer/trainer.py#L1261-L1269

### Motivation

Users can use `on_train_start`, `on_fit_start`, or `setup` for the same purpose.

### Additional Context

These hooks were originally added in https://github.com/PyTorchLightning/pytorch-lightning/pull/2850

I see about 30 usages of each hook using grep.app: 
https://www.grep.app/search?q=on_pretrain_routine_start 
https://www.grep.app/search?q=on_pretrain_routine_end

Part of #7740
Carries forward discussion #9043 on these hooks

cc @borda @tchaton @rohitgr7 @carmocca @awaelchli @ninginthecloud @daniellepintz

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #10984

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: tests/trainer/test_data_loading.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def __init__(self, num_workers):
     def train_dataloader(self):
         return DataLoader(RandomDataset(32, 64), num_workers=self.num_workers)
 
-    def on_pretrain_routine_start(self):
+    def setup(self, stage=None):
```
Comment (awaelchli) [inline tests/trainer/test_data_loading.py:107]: `on_fit_start` would also work here, right?
It makes more sense because `on_pretrain_routine_start` also only ran for fitting.

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_restore.py [hunk 2]
```diff
@@ -622,14 +622,14 @@ def on_pretrain_routine_end(self):
...
-            self.on_pretrain_routine_end_called = True
+            self.on_train_start_called = True
 
     # new model
     model = CustomModel()
 
     # fit new model which should load hpc weights
     new_trainer.fit(model, datamodule=dm)
-    assert model.on_pretrain_routine_end_called
+    assert model.on_train_start_called
```
Comment (akihironitta) [inline tests/models/test_restore.py:632]: The test case here is failing. Any idea?

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by daniellepintz — APPROVED** [review]

File: tests/models/test_restore.py (hunk 2 above)
Comment (ananthsub) [reply tests/models/test_restore.py:632]: I'm having a hard time. I don't really understand what this test is trying to do

Comment (ananthsub) [reply tests/models/test_restore.py]: it looks like these tests predated `trainer.validate` so there's no need to write it this way anymore
